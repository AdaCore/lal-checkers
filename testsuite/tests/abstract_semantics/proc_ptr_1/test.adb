procedure Test is
   X : Integer;
   Y : Integer;
   B : Boolean;

   procedure F is
   begin
      X := 12;
   end F;

   procedure G is
   begin
      X := 14;
      Y := 42;
   end G;

   P : access procedure;
begin
   if B then
      P := F'Access;
   else
      P := G'Access;
   end if;
end Test;
