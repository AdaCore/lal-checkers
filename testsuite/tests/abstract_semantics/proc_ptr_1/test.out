{
  "always": {
    "assign0": [
      {
        "trace:": [
          "assign0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}"
        }
      }
    ], 
    "assign1": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]', 2: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}"
        }
      }
    ], 
    "assign2": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]', 2: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}"
        }
      }
    ], 
    "assign3": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]', 2: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}"
        }
      }
    ], 
    "assign4": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "assume0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]', 2: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{True}", 
          "P": "{Subprogram <SubpBody [\"F\"] test.adb:6:4-9:10> capturing (0x00000001)}"
        }
      }
    ], 
    "assign5": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign5", 
          "assume1", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]', 2: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{False}", 
          "P": "{Subprogram <SubpBody [\"G\"] test.adb:11:4-15:10> capturing (0x00000001, 0x00000002)}"
        }
      }
    ], 
    "assume0": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assume0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]', 2: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{True}", 
          "P": "{[all-path]}"
        }
      }
    ], 
    "assume1": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assume1", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]', 2: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{False}", 
          "P": "{[all-path]}"
        }
      }
    ], 
    "split_join0": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "assign5", 
          "assume0", 
          "assume1", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]', 2: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{False, True}", 
          "P": "{Subprogram <SubpBody [\"F\"] test.adb:6:4-9:10> capturing (0x00000001), Subprogram <SubpBody [\"G\"] test.adb:11:4-15:10> capturing (0x00000001, 0x00000002)}"
        }
      }
    ], 
    "start0": [
      {
        "trace:": [
          "start0"
        ], 
        "values": {
          "$stack": "({}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}"
        }
      }
    ]
  }, 
  "le_t_eq_v": {
    "assign0": [
      {
        "trace:": [
          "assign0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}"
        }
      }
    ], 
    "assign1": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]', 2: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}"
        }
      }
    ], 
    "assign2": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]', 2: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}"
        }
      }
    ], 
    "assign3": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]', 2: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}"
        }
      }
    ], 
    "assign4": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "assume0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]', 2: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{True}", 
          "P": "{Subprogram <SubpBody [\"F\"] test.adb:6:4-9:10> capturing (0x00000001)}"
        }
      }
    ], 
    "assign5": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign5", 
          "assume1", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]', 2: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{False}", 
          "P": "{Subprogram <SubpBody [\"G\"] test.adb:11:4-15:10> capturing (0x00000001, 0x00000002)}"
        }
      }
    ], 
    "assume0": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assume0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]', 2: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{True}", 
          "P": "{[all-path]}"
        }
      }
    ], 
    "assume1": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assume1", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]', 2: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{False}", 
          "P": "{[all-path]}"
        }
      }
    ], 
    "split_join0": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "assume0", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]', 2: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{True}", 
          "P": "{Subprogram <SubpBody [\"F\"] test.adb:6:4-9:10> capturing (0x00000001)}"
        }
      }, 
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign5", 
          "assume1", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[-2147483648, 2147483647]', 2: '[-2147483648, 2147483647]'}, 0)", 
          "B": "{False}", 
          "P": "{Subprogram <SubpBody [\"G\"] test.adb:11:4-15:10> capturing (0x00000001, 0x00000002)}"
        }
      }
    ], 
    "start0": [
      {
        "trace:": [
          "start0"
        ], 
        "values": {
          "$stack": "({}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}"
        }
      }
    ]
  }
}
